{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d1ae905-3a53-40f5-99c6-21cd7bba038e",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(ggplot2)\n",
    "library(matrixStats)\n",
    "library(ggsci)\n",
    "library(magrittr) \n",
    "library(ggrepel)\n",
    "library(dplyr)\n",
    "library(ComplexHeatmap)\n",
    "library(ggpubr)\n",
    "library(circlize)\n",
    "library(hrbrthemes)\n",
    "\n",
    "confidence_interval_upper <- function(vector, interval=0.95) {\n",
    "  # Standard deviation of sample\n",
    "  vec_sd <- sd(vector, na.rm = TRUE)\n",
    "  # Sample size\n",
    "  n <- length(vector)\n",
    "  # Median of sample\n",
    "  vec_mean <- mean(vector, na.rm = TRUE)\n",
    "  # Error according to t distribution\n",
    "  error <- qt((interval + 1)/2, df = n - 1) * vec_sd / sqrt(n)\n",
    "  # Confidence interval as a vector\n",
    "  result <- vec_mean + error\n",
    "  return(result)\n",
    "}\n",
    "confidence_interval_lower <- function(vector, interval=0.95) {\n",
    "  # Standard deviation of sample\n",
    "  vec_sd <- sd(vector, na.rm = TRUE)\n",
    "  # Sample size\n",
    "  n <- length(vector)\n",
    "  # Median of sample\n",
    "  vec_mean <- mean(vector, na.rm = TRUE)\n",
    "  # Error according to t distribution\n",
    "  error <- qt((interval + 1)/2, df = n - 1) * vec_sd / sqrt(n)\n",
    "  # Confidence interval as a vector\n",
    "  result <- vec_mean - error\n",
    "  return(result)\n",
    "}\n",
    "createEmptyDf = function( nrow, ncol, colnames = c() ){\n",
    "  if( missing( ncol ) && length( colnames ) > 0 ){\n",
    "    ncol = length( colnames )\n",
    "  }\n",
    "  data.frame( matrix( vector(), nrow, ncol, dimnames = list( c(), colnames ) ) )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca733577-14bd-4184-9c8f-0dbebd5801c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/Users/inamojun/TMDU-LR_isoform_atlas/Figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff75f75f-607f-4fea-be11-0a06d44d5282",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "load(\"../data/data_Fig02.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "273c40ab-fc1a-4f7d-873d-7d8e5b0de9be",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6af88406-06e2-42b1-96ed-74ea616b5861",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(mat_var) # top 5000 high variable isoforms among cell types after normalizing by Read Per Million\n",
    "dim(mat_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45d7d3dc-ea9e-4197-86b1-b1500649578d",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(mat) # specific isoforms after normalizing by Read Per Million\n",
    "dim(mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0c64a30-cca3-4dd7-909d-479bbac0d26a",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(col_lab) # with label\n",
    "dim(col_lab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cebac3de-f197-4c33-bc04-e4c5c220d693",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(elavl) # ELAVL1 RIP-seq peaks and their region, isoform expression of RNA-seq derived from relevant cells (LCL, GEUVADIS)\n",
    "dim(elavl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdd4a6a5-f3f6-4473-a390-c9bbf51be82e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hirarchial clustering by expression correlation among cell types using the top 5000 variable isoforms\n",
    "ha = HeatmapAnnotation(cell = factor(colnames(mat_var), levels = colnames(mat_var)),\n",
    "                       col = list(cell = c(\"NaiveCD4\" = \"chartreuse3\",\"Th1\" = \"chartreuse3\",\"Th2\" = \"chartreuse3\",\"Th17\" = \"chartreuse3\",\"Tfh\" = \"chartreuse3\",\"Fr.I nTreg\" = \"chartreuse3\",\"Fr.II aTreg\" = \"chartreuse3\",\"Fr.III non-Treg\" = \"chartreuse3\",\"LAG3 Treg\" = \"chartreuse3\",\"Memory CD4\" = \"chartreuse3\",\"Thx\" = \"chartreuse3\",\n",
    "                                           \"NaiveCD8\" = \"steelblue3\",\"Memory CD8\" = \"steelblue3\",\"CM CD8\" = \"steelblue3\",\"EM CD8\" = \"steelblue3\",\n",
    "                                           \"NaiveB\" = \"chocolate1\",\"USMB\" = \"chocolate1\",\"SMB\" = \"chocolate1\",\"DNB\" = \"chocolate1\",\"plasmablast\" = \"chocolate1\",\n",
    "                                           \"pDC\" = \"wheat3\",\"mDC\" = \"wheat3\",\n",
    "                                           \"NK\" = \"grey45\",\n",
    "                                           \"CL Mono\" = \"brown1\",\"NC Mono\" = \"brown1\",\"Int Mono\" = \"brown1\",\"CD16p Mono\" = \"brown1\",\n",
    "                                           \"PBMC\" = \"black\",\n",
    "                                           \"Neu\" = \"darkmagenta\")),\n",
    "                       show_annotation_name = FALSE)\n",
    "                \n",
    "options(repr.plot.width=8, repr.plot.height=6)\n",
    "Heatmap(cor(mat_var), name = \"cor\", \n",
    "        top_annotation = ha,\n",
    "        col = circlize::colorRamp2(c(-1, 0, 1), c(\"green\", \"white\", \"red\")), \n",
    "        show_row_names = TRUE, show_column_names = TRUE, show_row_dend = FALSE, show_column_dend = FALSE, \n",
    "        column_title = \"pairwise correlation between cell-types\",\n",
    "        heatmap_legend_param = list(title = \"Correlation\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41497a2c-3759-451d-a507-d506a118ce91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specific isoforms\n",
    "\n",
    "options(repr.plot.width=16, repr.plot.height=8)\n",
    "# gwas_genes: efotraits_EFO_0000540-associations-2022-06-14.csv\n",
    "\n",
    "ht_list = \n",
    "  HeatmapAnnotation(\n",
    "    label = anno_mark(at = which(col_lab$immune_or_TF & col_lab$associated_gene %in% gwas_genes), \n",
    "                      labels = col_lab[col_lab$immune_or_TF & col_lab$associated_gene %in% gwas_genes, ]$id, \n",
    "                      side = \"top\",\n",
    "                      labels_gp = gpar(fontsize = 9\n",
    "                                       # col = rep(c(\"palegreen\",\"palegreen3\",\"palegreen4\",\"chartreuse\",\"chartreuse3\",\"chartreuse4\",\"green1\",\"green3\",\"olivedrab1\",\"olivedrab2\",\"olivedrab3\",\n",
    "                                       #             \"steelblue1\",\"steelblue2\",\"steelblue3\",\"steelblue4\",\n",
    "                                       #             \"wheat1\",\"wheat2\",\"wheat3\",\n",
    "                                       #             \"grey90\",\n",
    "                                       #             \"brown1\",\"brown2\",\"brown3\",\"brown4\",\n",
    "                                       #             \"black\",\n",
    "                                       #             \"darkmagenta\"),each=3)\n",
    "                      ), \n",
    "                      padding = 0,\n",
    "                      link_width = unit(0.01, \"mm\"),\n",
    "                      link_height = unit(10, \"mm\")\n",
    "    )) %v%\n",
    "  Heatmap(scale(t(mat)), \n",
    "          cluster_columns = FALSE, show_column_names = FALSE, show_column_dend = FALSE,\n",
    "          cluster_rows = FALSE, \n",
    "          row_names_gp = gpar(fontsize = 12, axis = 45),\n",
    "          heatmap_legend_param = list(\n",
    "            # at = c(-2, 0, 2),\n",
    "            # labels = c(\"low\", \"zero\", \"high\"),\n",
    "            title = \"scaled expression value\",\n",
    "            legend_height = unit(4, \"cm\"),\n",
    "            direction = \"horizontal\",\n",
    "            title_position = \"topcenter\") \n",
    "  ) \n",
    "draw(column_title = paste0(\"specific expressed isoforms [n=\",nrow(mat),\"]\"), ht_list, heatmap_legend_side=\"bottom\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb702f61-116a-44d4-941d-a1219f8b1cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=12, repr.plot.height=8)\n",
    "\n",
    "cell_ord = c(\"NaiveCD4\",\"Th1\",\"Th2\",\"Th17\",\"Tfh\",\"Fra1\",\"Fra2.aTreg\",\"Fra3\",\"LAG3Treg\",\"MemoryCD4\",\"Thx\",\n",
    "             \"NaiveCD8\",\"CD8effector\",\"CD8centralmem\",\"CD8effectormem\",\n",
    "             \"NaiveB\",\"unswmemoryB\",\"swmemoryB\",\"DNB\",\n",
    "             \"plasmablast\",\"plasmacytoidDC\",\"myeloidDC\",\n",
    "             \"NK\",\n",
    "             \"monocyteCD16\",\"monocyteCD16minus\",\"nonclassicalMonocyte\",\"intermediateMonocyte\",\n",
    "             \"PBMC\",\n",
    "             \"Neutrophil\")\n",
    "ha = HeatmapAnnotation(specific_cell = factor(col_lab_dice$specific_cell_LR, levels = intersect(cell_ord,col_lab_dice$specific_cell_LR)),\n",
    "                       col = list(specific_cell = c(\"NaiveB\" = \"chocolate1\",\"unswmemoryB\" = \"chocolate2\",\"swmemoryB\" = \"chocolate3\",\"DNB\" = \"chocolate4\",\n",
    "                                                    \"NaiveCD4\" = \"palegreen\",\"Th1\" = \"palegreen3\",\"Th2\" = \"palegreen4\",\"Th17\" = \"chartreuse\",\"Tfh\" = \"chartreuse3\",\"Fra1\" = \"chartreuse4\",\"Fra2.aTreg\" = \"green1\",\"Fra3\" = \"green3\",\"LAG3Treg\" = \"olivedrab1\",\"MemoryCD4\" = \"olivedrab2\",\"Thx\" = \"olivedrab3\",\n",
    "                                                    \"NaiveCD8\" = \"steelblue1\",\"CD8effector\" = \"steelblue2\",\"CD8centralmem\" = \"steelblue3\",\"CD8effectormem\" = \"steelblue4\",\n",
    "                                                    \"plasmablast\" = \"wheat1\",\"plasmacytoidDC\" = \"wheat2\",\"myeloidDC\" = \"wheat3\",\n",
    "                                                    \"NK\" = \"grey45\",\n",
    "                                                    \"monocyteCD16\" = \"brown1\",\"monocyteCD16minus\" = \"brown2\",\"nonclassicalMonocyte\" = \"brown3\",\"intermediateMonocyte\" = \"brown4\",\n",
    "                                                    \"PBMC\" = \"black\",\n",
    "                                                    \"Neutrophil\" = \"darkmagenta\")),\n",
    "                       annotation_name_side = \"left\")\n",
    "ht_list = \n",
    "  Heatmap(scale(t(dice)), \n",
    "          top_annotation = ha, \n",
    "          cluster_columns = FALSE, show_column_names = FALSE, show_column_dend = FALSE,\n",
    "          cluster_rows = FALSE, show_row_names = FALSE, show_row_dend = FALSE, \n",
    "          row_names_gp = gpar(fontsize = 10, axis = 45),\n",
    "          row_split = \n",
    "            IID_d$cell %>%\n",
    "            as.data.frame() %>%\n",
    "            magrittr::set_colnames(\"cell\") %>%\n",
    "            dplyr::mutate(group = dplyr::case_when(\n",
    "              cell == \"B_NAIVE\" ~ \"B\",\n",
    "              cell == \"CD4_NAIVE\" ~ \"CD4T\",cell == \"CD4_N_STIM\" ~ \"CD4T\",cell == \"TH1\" ~ \"CD4T\",cell == \"TH17\" ~ \"CD4T\",cell == \"TH2\" ~ \"CD4T\",cell == \"TH1-17\" ~ \"CD4T\",cell == \"TFH\" ~ \"CD4T\",cell == \"TREG_MEMORY\" ~ \"CD4T\",cell == \"TREG_NAIVE\" ~ \"CD4T\",\n",
    "              cell == \"CD8_NAIVE\" ~ \"CD8T\",cell == \"CD8_N_STIM\" ~ \"CD8T\",\n",
    "              cell == \"NONCLASSICAL_MONOCYTES\" ~ \"Mono\",cell == \"CLASSICAL_MONOCYTES\" ~ \"Mono\",\n",
    "              cell == \"NK_CD16POS\" ~ \"NK\"\n",
    "            )) %>%\n",
    "            .$group\n",
    "          , \n",
    "          row_title = \"DICE cell groups\",\n",
    "          column_split = col_lab_dice$specific_cell_LR %>%\n",
    "            as.data.frame() %>%\n",
    "            magrittr::set_colnames(\"cell\") %>%\n",
    "            dplyr::mutate(group = dplyr::case_when(\n",
    "              cell == \"NaiveB\" ~ \"B\",cell == \"unswmemoryB\" ~ \"B\",cell == \"swmemoryB\" ~ \"B\",cell == \"DNB\" ~ \"B\",cell == \"plasmablast\" ~ \"B\",\n",
    "              cell == \"NaiveCD4\" ~ \"CD4T\",cell == \"Th1\" ~ \"CD4T\",cell == \"Th2\" ~ \"CD4T\",cell == \"Th17\" ~ \"CD4T\",cell == \"Tfh\" ~ \"CD4T\",cell == \"Fra1\" ~ \"CD4T\",cell == \"Fra2.aTreg\" ~ \"CD4T\",cell == \"Fra3\" ~ \"CD4T\",cell == \"LAG3Treg\" ~ \"CD4T\",cell == \"MemoryCD4\" ~ \"CD4T\",cell == \"Thx\" ~ \"CD4T\",\n",
    "              cell == \"NaiveCD8\" ~ \"CD8T\",cell == \"CD8effector\" ~ \"CD8T\",cell == \"CD8centralmem\" ~ \"CD8T\",cell == \"CD8effectormem\" ~ \"CD8T\",\n",
    "              cell == \"monocyteCD16\" ~ \"Mono\",cell == \"nonclassicalMonocyte\" ~ \"Mono\",cell == \"intermediateMonocyte\" ~ \"Mono\",cell == \"monocyteCD16minus\" ~ \"Mono\",\n",
    "              cell == \"NK\" ~ \"NK\"\n",
    "            )) %>%\n",
    "            .$group,\n",
    "          column_title = \"LR cell groups\",\n",
    "          heatmap_legend_param = list(\n",
    "            # at = c(-2, 0, 2),\n",
    "            # labels = c(\"low\", \"zero\", \"high\"),\n",
    "            title = \"scaled expression value\",\n",
    "            legend_height = unit(4, \"cm\")) \n",
    "  ) +\n",
    "  Heatmap(IID_d$cell, \n",
    "          name = \"DICE\", \n",
    "          col = c(\"B_NAIVE\" = \"darksalmon\",\n",
    "                  \"CD4_NAIVE\" = \"aquamarine1\",\"CD4_N_STIM\" = \"olivedrab3\",\"TH1\" = \"darkseagreen1\",\"TH17\" = \"darkseagreen2\",\"TH2\" = \"darkseagreen3\",\"TH1-17\" = \"darkseagreen4\",\"TFH\" = \"chartreuse3\",\"TREG_MEMORY\" = \"green1\",\"TREG_NAIVE\" = \"green3\",\n",
    "                  \"CD8_NAIVE\" = \"deepskyblue\",\"CD8_N_STIM\" = \"deepskyblue3\",\n",
    "                  \"NONCLASSICAL_MONOCYTES\" = \"firebrick1\",\"CLASSICAL_MONOCYTES\" = \"firebrick3\",\n",
    "                  \"NK_CD16POS\" = \"grey45\"),\n",
    "          width = unit(5, \"mm\"))\n",
    "draw(column_title = paste0(\"expression pattern of specific isoforms matched with relevant cell types in DICE [n=\",nrow(dice),\" (\",format(100*nrow(dice)/nrow(mat),digits = 4),\"%)]\"), \n",
    "     ht_list, \n",
    "     merge_legends = TRUE,\n",
    "     heatmap_legend_side=\"right\")\n",
    "\n",
    "\n",
    "# Are specifically expressed cell-types in isoform atlas matched with other datasets?\n",
    "## permutation test using DICE\n",
    "### The permutation test is performed by shuffling of specific_cell_LR label 1000 times to statistically test whether the above-mentioned number of matches is large or not (whether the specific expressed cell in LR matches the DICE cell types by chance or not). (i.e., whether the specific expressed cells in LR are consistent with the DICE cell types rather than coincidental).\n",
    "### Result\n",
    "#### Cell types of 15% (391/2575) of isoforms specifically expressed in LR-PB29 matched concordant cell types in DICE\n",
    "#### According to the permutation test shuffled label of specifically expressed cell type (n=1000),                                    \n",
    "#### specifically expressed cell types of LR-PB29 and top cell types of DICE* matched significantly (p=0.004).\n",
    "##### * Pairs of cell-types\n",
    "##### B\n",
    "##### DICE: B_NAIVE \n",
    "##### LR-PB29: NaiveB, unswmemoryB, swmemoryB, DNB, plasmablast\n",
    "##### CD4T\n",
    "##### DICE: CD4_NAIVE, CD4_N_STIM, TH1, TH17, TH2, TH1-17, TFH, TREG_MEMORY, TREG_NAIVE\n",
    "##### LR-PB29: NaiveCD4, Th1, Th2, Th17, Tfh,Fra1.Treg, Fra2.aTreg, Fra3.Treg, LAG3.Treg, MemoryCD4, Thx\n",
    "##### CD8T\n",
    "##### DICE: CD8_NAIVE, CD8_N_STIM\n",
    "##### LR-PB29: NaiveCD8, CD8effector, CD8centralmem, CD8effectormem\n",
    "##### Monocyte\n",
    "##### DICE: CLASSICAL_MONOCYTES, NONCLASSICAL_MONOCYTES\n",
    "##### LR-PB29: monocyteCD16, monocyteCD16minus, nonclassicalMonocyte, intermediateMonocyte\n",
    "##### NK\n",
    "##### DICE: NK_CD16POS\n",
    "##### LR-PB29: NK\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2f6228c-f19b-4a25-bf79-2712904b5016",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare length between specific isoforms and non-specific isoforms\n",
    "options(repr.plot.width=8, repr.plot.height=4)\n",
    "# compare length\n",
    "variable = \"specificity_LR\"\n",
    "for (region in c(\"five_utr_length\",\"CDS_length\",\"three_utr_length\")){\n",
    "  lab = dplyr::case_when(\n",
    "    region == \"five_utr_length\" ~ \"5'UTR length\",\n",
    "    region == \"CDS_length\" ~ \"ORF length\",\n",
    "    region == \"three_utr_length\" ~ \"3'UTR length\"\n",
    "  )\n",
    "  g = out %>%\n",
    "    dplyr::mutate(strata = eval(parse(text = variable)),\n",
    "                  strata = dplyr::case_when(\n",
    "                    strata ~ \"specific\",\n",
    "                    !strata ~ \"non-specific\"\n",
    "                  )) %>%\n",
    "    ggboxplot(., x=\"strata\", y=region,\n",
    "              color = \"strata\",\n",
    "              outlier.shape = NA) + \n",
    "    coord_cartesian(ylim = quantile(eval(parse(text=paste0(\"out$\",region))), c(0, 0.9), na.rm=TRUE)) +\n",
    "    stat_compare_means(\n",
    "      label = \"p.signif\",\n",
    "      label.y = quantile(eval(parse(text=paste0(\"out$\",region))), 0.88, na.rm=TRUE),\n",
    "      label.x.npc = \"center\") + \n",
    "    theme_classic() +\n",
    "    ylab(lab) +\n",
    "    xlab(\"\") +\n",
    "    theme(strip.text.x=element_text(size=9, color=\"black\", face=\"bold\"),\n",
    "          strip.text.y=element_text(size=9, color=\"black\", face=\"bold\"),\n",
    "          legend.position = \"bottom\",\n",
    "          plot.title = element_text(size=20),\n",
    "          axis.title.x = element_text(size=20),\n",
    "          axis.title.y = element_text(size =20),\n",
    "          axis.text.y = element_text(size = 20),\n",
    "          axis.text.x = element_text(size = 20,angle = 45, hjust=1),\n",
    "          legend.text =  element_text(size = 20), \n",
    "          legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "          legend.title = element_text(size = 0, hjust = 0))\n",
    "  assign(paste0(\"g_\",region),g)\n",
    "}\n",
    "\n",
    "ggpubr::ggarrange(g_five_utr_length,g_CDS_length,g_three_utr_length,ncol=3,common.legend = TRUE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9dfb7eb-a0a2-4955-a669-6d69bd5fc717",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment of RBP binding motif by RBPmap\n",
    "## specific isoforms and others (LR cell-types, SR cell-types, SR stimulus)\n",
    "## translational efficiency top10 vs bottom10)\n",
    "\n",
    "length(unique(pval_rbp$RBP)) # total number of tested RBP\n",
    "\n",
    "# ELAVL1\n",
    "pval_rbp %>%\n",
    "  dplyr::filter(RBP==\"HuR\") %>%\n",
    "  dplyr::arrange(-CI_up)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "955f764b-68d6-4afb-a17d-4b4a97b8b8f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IGF2BP3\n",
    "## Insulin-like growth factor 2 mRNA-binding protein 3 is a protein that in humans is encoded by the IGF2BP3 gene.\n",
    "## The protein encoded by this gene is primarily found in the nucleolus, where it can bind to the 5' UTR of the insulin-like growth factor II leader 3 mRNA and may repress translation of insulin-like growth factor II during late development.\n",
    "pval_rbp %>%\n",
    "  dplyr::filter(RBP==\"IGF2BP3\") %>%\n",
    "  dplyr::filter(strata==\"te_rank\") %>%\n",
    "  dplyr::arrange(-CI_up)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf29fa5a-3165-42c5-9420-a52041bb0cea",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8, repr.plot.height=4)\n",
    "\n",
    "variable=\"specificity_LR\"\n",
    "genes=c(\"HuR\",InnateDB)\n",
    "# variable=\"specificity_SR_celltype\"\n",
    "# variable=\"specificity_SR_stim\"\n",
    "# variable=\"InnateDB\"\n",
    "# variable=\"te_rank\"\n",
    "variables=c(\"specificity_LR\",\"specificity_LRgroup\",\"specificity_SR_celltype\",\"specificity_SR_stim\",\"te_rank\")\n",
    "for (variable in variables){\n",
    "  print(variable)\n",
    "  mat = pval_rbp %>%\n",
    "    dplyr::filter(region != \"total\") %>%\n",
    "    dplyr::filter(strata == variable) %>%\n",
    "    dplyr::mutate(log_odds = log(odds),\n",
    "                  region = factor(region,levels=c(\"Intron\",\"3UTR\",\"CDS\",\"5UTR\",\"total\"))\n",
    "    ) %>%\n",
    "    dplyr::select(c(log_odds,region,RBP)) %>%\n",
    "    tidyr::pivot_wider(names_from = \"RBP\",\n",
    "                       values_from = \"log_odds\") %>%\n",
    "    as.data.frame() %>%\n",
    "    tibble::column_to_rownames(\"region\") %>%\n",
    "    as.matrix()\n",
    "  dim(mat)\n",
    "  lab = dplyr::case_when(\n",
    "    variable == \"specificity_LR\" ~ paste0(\"[specific vs non-specific]\"),\n",
    "    variable == \"specificity_LRgroup\" ~ paste0(\"[specific vs non-specific]\"),\n",
    "    variable == \"specificity_SR_celltype\" ~ paste0(\"[specific vs non-specific]\"),\n",
    "    variable == \"specificity_SR_stim\" ~ paste0(\"[specific vs non-specific]\"),\n",
    "    variable == \"te_rank\" ~ paste0(\"translational efficiency\\n[top10 vs bottom10]\"))\n",
    "  anno_label = pval_rbp %>%\n",
    "    dplyr::filter(region != \"total\") %>%\n",
    "    dplyr::filter(strata == variable) %>%\n",
    "    dplyr::mutate(sig = ifelse(CI_up<1 | CI_low>1, TRUE, FALSE),\n",
    "                  interest = ifelse(RBP %in% genes, TRUE, FALSE)) %>%\n",
    "    dplyr::group_by(RBP) %>%\n",
    "    dplyr::mutate(sig = ifelse(any(sig), TRUE, FALSE)) %>%\n",
    "    as.data.frame() %>%\n",
    "    dplyr::distinct(RBP,sig,interest) %>%\n",
    "    magrittr::set_rownames(.$RBP) %>%\n",
    "    .[colnames(mat),]\n",
    "  show_heatmap_legend = ifelse(variable == \"specificity_LR\", TRUE, FALSE)\n",
    "  mat = mat[c(\"5UTR\",\"CDS\",\"3UTR\"),]\n",
    "  ht_list = \n",
    "    Heatmap(mat, \n",
    "            top_annotation = HeatmapAnnotation(mark = anno_mark(at = which(anno_label$interest & anno_label$sig), \n",
    "                                                                labels = anno_label[anno_label$interest & anno_label$sig, ]$RBP, \n",
    "                                                                side = \"top\",\n",
    "                                                                labels_rot = 45,\n",
    "                                                                labels_gp = gpar(fontsize = 10), \n",
    "                                                                link_width = unit(5, \"mm\"), \n",
    "                                                                padding = unit(3, \"mm\")\n",
    "            )),\n",
    "            cluster_columns = TRUE, show_column_names = FALSE, show_column_dend = FALSE,\n",
    "            cluster_rows = FALSE, show_row_names = TRUE, show_row_dend = FALSE, \n",
    "            row_names_gp = gpar(fontsize = 10, axis = 0),\n",
    "            column_title = lab, \n",
    "            column_title_side = \"top\",\n",
    "            column_title_gp = gpar(fontsize = 10, axis = 45),\n",
    "            show_heatmap_legend = show_heatmap_legend,\n",
    "            heatmap_legend_param = list(\n",
    "              at = c(-2, 0, 2),\n",
    "              # labels = c(\"low\", \"zero\", \"high\"),\n",
    "              title = \"log(odds ratio)\",\n",
    "              legend_height = unit(4, \"cm\"),\n",
    "              legend_width = unit(3, \"cm\"),\n",
    "              legend_side = \"bottom\",\n",
    "              direction = \"horizontal\") , \n",
    "            width = unit(6, \"cm\"),\n",
    "            height = unit(2, \"cm\")\n",
    "    )\n",
    "  assign(paste0(\"ht_list_\",variable),ht_list)\n",
    "}\n",
    "\n",
    "draw(column_title = \"enrichment of RBP binding motifs\", \n",
    "     column_title_gp = gpar(fontsize = 10, axis = 45),\n",
    "     ht_list_specificity_LR + \n",
    "     ht_list_specificity_LRgroup + \n",
    "     ht_list_specificity_SR_celltype + \n",
    "     ht_list_specificity_SR_stim + \n",
    "     ht_list_te_rank , \n",
    "     merge_legends = TRUE,\n",
    "     heatmap_legend_side=\"bottom\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6b31d5a-bbd2-4e16-87fd-af1076d472ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=4)\n",
    "\n",
    "## correlation with isoform expression\n",
    "### max of GEUVADIS\n",
    "elavl %>%\n",
    "  dplyr::group_by(isoform) %>%\n",
    "  dplyr::mutate(isoform_score = sum(pileup)) %>%\n",
    "  dplyr::arrange(isoform) %>%\n",
    "  as.data.frame() %>%\n",
    "  dplyr::distinct(isoform,isoform_score,log_tpm_max) %>%\n",
    "  ggscatter(., x = \"isoform_score\", y = \"log_tpm_max\",\n",
    "            color = \"black\", size = 1, alpha = 0.1, \n",
    "            add = \"reg.line\",  \n",
    "            add.params = list(color = \"blue\", fill = \"lightgray\"),\n",
    "            conf.int = TRUE, \n",
    "            cor.coef = TRUE,\n",
    "            cor.coeff.args = list(method = \"pearson\", label.x.npc = \"center\", label.y.npc = \"top\", label.sep = \"\\n\", color=\"red\")\n",
    "  ) +\n",
    "  scale_x_log10()\n",
    "\n",
    "### median of GEUVADIS\n",
    "elavl %>%\n",
    "  dplyr::group_by(isoform) %>%\n",
    "  dplyr::mutate(isoform_score = sum(pileup)) %>%\n",
    "  dplyr::arrange(isoform) %>%\n",
    "  as.data.frame() %>%\n",
    "  dplyr::distinct(isoform,isoform_score,log_tpm_med) %>%\n",
    "  ggscatter(., x = \"isoform_score\", y = \"log_tpm_med\",\n",
    "            color = \"black\", size = 1, alpha = 0.1, \n",
    "            add = \"reg.line\",  \n",
    "            add.params = list(color = \"blue\", fill = \"lightgray\"),\n",
    "            conf.int = TRUE, \n",
    "            cor.coef = TRUE,\n",
    "            cor.coeff.args = list(method = \"pearson\", label.x.npc = \"center\", label.y.npc = \"top\", label.sep = \"\\n\", color=\"red\")\n",
    "  ) +\n",
    "  scale_x_log10()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f867dc-29b2-41d5-ad4c-abdace6d91d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# summary of peak score\n",
    "elavl %>%\n",
    "  # dplyr::filter(log_tpm_med>0) %>%\n",
    "  dplyr::distinct(isoform,region,peak_id,.keep_all = TRUE) %>%\n",
    "  dplyr::mutate(adj_count = pileup/(log_tpm_med_sum+1),\n",
    "                region = factor(region,levels=c(\"5UTR\",\"CDS\",\"3UTR\",\"Intron\"))) %>%\n",
    "  dplyr::group_by(isoform,region) %>%\n",
    "  dplyr::mutate(adj_count_sum = log(sum(adj_count)+1)) %>%\n",
    "  dplyr::arrange(isoform,region) %>%\n",
    "  as.data.frame() %>%\n",
    "  dplyr::distinct(isoform,region,.keep_all = TRUE) %>%\n",
    "  .$adj_count_sum %>%\n",
    "  summary()\n",
    "\n",
    "options(repr.plot.width=4, repr.plot.height=4)\n",
    "elavl %>%\n",
    "  # dplyr::filter(log_tpm_med>0) %>%\n",
    "  dplyr::distinct(isoform,region,peak_id,.keep_all = TRUE) %>%\n",
    "  dplyr::mutate(adj_count = pileup/(log_tpm_med_sum+1),\n",
    "                region = factor(region,levels=c(\"5UTR\",\"CDS\",\"3UTR\",\"Intron\"))) %>%\n",
    "  dplyr::group_by(isoform,region) %>%\n",
    "  dplyr::mutate(adj_count_sum = log(sum(adj_count)+1)) %>%\n",
    "  dplyr::arrange(isoform,region) %>%\n",
    "  as.data.frame() %>%\n",
    "  dplyr::distinct(isoform,region,.keep_all = TRUE) %>%\n",
    "  .$adj_count_sum %>%\n",
    "  hist(., breaks = 50, main = \"histgram of peak score\", xlab = \"peak score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21d90179-c6f2-40d9-944e-bd682bfb3395",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare peak scores according to specificity and translational efficiency\n",
    "for (variable in c(\"specificity_LR\",\"specificity_SR_celltype\",\"specificity_SR_stim\",\"te_rank\")){\n",
    "  num = grep(variable,c(\"specificity_LR\",\"specificity_SR_celltype\",\"specificity_SR_stim\",\"te_rank\"))+1\n",
    "  g = elavl %>%\n",
    "    # dplyr::filter(log_tpm_med>0) %>%\n",
    "    dplyr::distinct(isoform,region,peak_id,.keep_all = TRUE) %>%\n",
    "    dplyr::mutate(adj_count = pileup/(log_tpm_med_sum+1),\n",
    "                  region = factor(region,levels=c(\"5UTR\",\"CDS\",\"3UTR\",\"Intron\"))) %>%\n",
    "    dplyr::group_by(isoform,region) %>%\n",
    "    dplyr::mutate(adj_count_sum = log(sum(adj_count)+1)) %>%\n",
    "    dplyr::arrange(isoform,region) %>%\n",
    "    as.data.frame() %>%\n",
    "    dplyr::distinct(isoform,region,.keep_all = TRUE) %>%\n",
    "    dplyr::filter(!is.na(eval(parse(text=variable)))) %>%\n",
    "    ggplot(., aes(x=region, y=adj_count_sum, fill = eval(parse(text=variable)))) + \n",
    "    geom_violin(adjust=3, position = position_dodge(width = 0.8)) +\n",
    "    geom_boxplot(width=0.1, color=\"black\", na.rm = TRUE, outlier.shape = NA, position = position_dodge(width = 0.8)) +\n",
    "    stat_compare_means(label = \"p.signif\",\n",
    "                       label.x.npc = \"center\",\n",
    "                       symnum.args = list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), symbols = c(\"****\", \"***\", \"**\", \"*\", \"\"))) +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"peak score\") +\n",
    "    labs(fill=variable) +\n",
    "    theme_minimal() +\n",
    "    theme(strip.text.x=element_text(size=9, color=\"black\", face=\"bold\"),\n",
    "          strip.text.y=element_text(size=9, color=\"black\", face=\"bold\"),\n",
    "          panel.grid=element_blank(),\n",
    "          legend.position = \"bottom\",\n",
    "          plot.title = element_text(size=10),\n",
    "          axis.title.x = element_text(size=10),\n",
    "          axis.title.y = element_text(size =10),\n",
    "          axis.text.y = element_text(size = 10),\n",
    "          axis.text.x = element_text(size = 10),\n",
    "          legend.text =  element_text(size = 9), \n",
    "          legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "          legend.title = element_text(size = 9, hjust = 0))\n",
    "  plot(g)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4087eadd-6c79-425a-8e58-caf3acf78e32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
