{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54693e4f-f0e7-4d71-971f-40667ea3c901",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(data.table))\n",
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(RColorBrewer))\n",
    "suppressMessages(library(ggsci))\n",
    "suppressMessages(library(magrittr))\n",
    "suppressMessages(library(ggrepel))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(ggalluvial))\n",
    "suppressMessages(library(ggpubr))\n",
    "options(warn=-1) # use options(warn=0) To turn warnings back on\n",
    "options(dplyr.summarise.inform = FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16559a77-9fa8-41e3-b8bf-75845a1efbb2",
   "metadata": {},
   "source": [
    "# Figure 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65ecb888-a39d-441e-8846-616532243cdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/Users/inamojun/TMDU-LR_isoform_atlas/Figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "002e42b4-3b25-4207-a913-9a49e1e7e646",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load data\n",
    "load(\"../data/data_Fig01.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2e9d0ca-cad1-4d1b-89c1-45a34c1b1e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# head(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40e3ed64-266e-491e-8660-a26b8e41f2dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# head(gencode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3010fa-c840-4bf8-8572-22f353004386",
   "metadata": {},
   "outputs": [],
   "source": [
    "# head(lr_suppa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c1ca10-b24d-47e2-bc2d-407e21679a2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# head(gen_suppa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b43568e7-246f-4be1-9b72-e672f7be9639",
   "metadata": {},
   "outputs": [],
   "source": [
    "# head(AS_susceptibility_pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3075eb5-4e6b-4fd5-aad7-ab2f5063f442",
   "metadata": {},
   "outputs": [],
   "source": [
    "nrow(out) # the total number of isoforms\n",
    "length(unique(out$associated_gene)) # the total number of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df14ddd-fe97-425d-8f67-3c96f0a887c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare the length of transcripts with those of GENCODE38\n",
    "# summary(out$transcript_length)\n",
    "# summary(gencode$transcript_length)\n",
    "\n",
    "wilcox.test(log(gencode$transcript_length),log(out$transcript_length))\n",
    "\n",
    "options(repr.plot.width=8, repr.plot.height=6)\n",
    "fig1b = rbind(\n",
    "  out[,c(\"isoform\",\"transcript_length\",\"associated_gene\")] %>%\n",
    "    dplyr::mutate(group = \"Isoform Atlas\"),\n",
    "  gencode[,c(\"isoform\",\"transcript_length\",\"associated_gene\")] %>%\n",
    "    dplyr::mutate(group = \"GENCODE38\")\n",
    ") %>% \n",
    "  ggplot(., aes(x = log10(transcript_length), fill = group)) + \n",
    "  geom_histogram(position = \"identity\", alpha = 0.8,binwidth = 0.1) + \n",
    "  # ggtitle(\"histgram of transcript length\") +\n",
    "  ylab(paste0(\"frequency\")) +\n",
    "  xlab(expression(paste({Log[10]}, \"(transcript length)\", sep=\"\"))) +\n",
    "  theme_classic() +\n",
    "  scale_fill_npg() +\n",
    "  theme(strip.text.x=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        strip.text.y=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=20),\n",
    "        axis.title.x = element_text(size=15),\n",
    "        axis.title.y = element_text(size =15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        axis.text.x = element_text(size = 15),\n",
    "        legend.text =  element_text(size = 10), \n",
    "        legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "        legend.title = element_text(size = 0, hjust = 0))\n",
    "\n",
    "fig1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d899bc-3fd6-4d27-956c-6c252ea6f792",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=9, repr.plot.height=4)\n",
    "\n",
    "# genes overlapped with GENCODE38\n",
    "x <- list(\n",
    "  `Isoform Atlas\\n[predicted coding]` = out[out$coding,]$associated_gene,\n",
    "  `GENCODE38\\n[coding]` = gencode[gencode$gene_type==\"protein_coding\",]$associated_gene\n",
    ")\n",
    "g1 = ggvenn::ggvenn(x, \n",
    "               show_elements = FALSE, show_percentage = TRUE, \n",
    "               digits = 1, fill_color = c(\"yellow\", \"green\"), \n",
    "               fill_alpha = 0.5, stroke_color = \"white\", stroke_alpha = 0.5, \n",
    "               stroke_size = 0, stroke_linetype = \"solid\", set_name_color = \"black\", \n",
    "               set_name_size = 4, text_color = \"black\", text_size = 3.5, label_sep = \",\")\n",
    "x <- list(\n",
    "  `Isoform Atlas\\n[lncRNA]` = out[out$gene_biotype==\"lncRNA\" | out$gene_biotype==\"lincRNA\",]$associated_gene,\n",
    "  `GENCODE38\\n[lncRNA]` = gencode[gencode$gene_type==\"lncRNA\" | gencode$gene_type==\"lincRNA\",]$associated_gene\n",
    ")\n",
    "g2 = ggvenn::ggvenn(x, \n",
    "               show_elements = FALSE, show_percentage = TRUE, \n",
    "               digits = 1, fill_color = c(\"yellow\", \"green\"), \n",
    "               fill_alpha = 0.5, stroke_color = \"white\", stroke_alpha = 0.5, \n",
    "               stroke_size = 0, stroke_linetype = \"solid\", set_name_color = \"black\", \n",
    "               set_name_size = 4, text_color = \"black\", text_size = 3.5, label_sep = \",\") \n",
    "\n",
    "## isoform level\n",
    "## isoforms overlapped with GENCODE38\n",
    "x <- list(\n",
    "  `Isoform Atlas\\n[isoform]` = out$isoform,\n",
    "  `GENCODE38\\n[isoform]` = gencode$isoform\n",
    ")\n",
    "\n",
    "\n",
    "g3 = ggvenn::ggvenn(x, \n",
    "               show_elements = FALSE, show_percentage = TRUE, \n",
    "               digits = 1, fill_color = c(\"yellow\", \"green\"), \n",
    "               fill_alpha = 0.5, stroke_color = \"white\", stroke_alpha = 0.5, \n",
    "               stroke_size = 0, stroke_linetype = \"solid\", set_name_color = \"black\", \n",
    "               set_name_size = 4, text_color = \"black\", text_size = 3.5, label_sep = \",\") \n",
    "\n",
    "# options(repr.plot.width=9, repr.plot.height=4)\n",
    "# ggpubr::ggarrange(g1,g2,g3, ncol=3)\n",
    "options(repr.plot.width=3, repr.plot.height=6)\n",
    "fig1c = ggpubr::ggarrange(g1,g3, ncol=1)\n",
    "\n",
    "fig1c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120a4ce2-f364-4597-a274-5a116cdea96e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specific coding-genes in isoform atlas\n",
    "# setdiff(out[out$coding,]$associated_gene,\n",
    "#         gencode[gencode$gene_type==\"protein_coding\",]$associated_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c509069c-9623-4152-9681-bd28c0645d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specific lncRNA in Isoform Atlas\n",
    "# setdiff(out[out$gene_biotype==\"lncRNA\",]$associated_gene,\n",
    "#         gencode[gencode$gene_type==\"lncRNA\",]$associated_gene)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6a17c6c-cc2e-46d1-bb4a-a63616e53f20",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=3)\n",
    "\n",
    "# the counts of isoforms per gene\n",
    "\n",
    "fig1d = rbind(out %>%\n",
    "        # dplyr::filter(gene_biotype == \"protein_coding\") %>%\n",
    "        dplyr::group_by(associated_gene) %>%\n",
    "        dplyr::summarise(count = dplyr::n()) %>%\n",
    "        dplyr::arrange(count) %>%\n",
    "        as.data.frame() %>%\n",
    "        dplyr::group_by(count) %>%\n",
    "        dplyr::summarise(isoform_count = dplyr::n()) %>%\n",
    "        as.data.frame() %>%\n",
    "        dplyr::mutate(count = ifelse(count>=10, \"10<=\", count),\n",
    "                     count = factor(count,levels = c(\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10<=\"))) %>%\n",
    "        dplyr::group_by(count) %>%\n",
    "        dplyr::mutate(isoform_count = sum(isoform_count),\n",
    "                      proportion = 100*isoform_count/length(unique(out[out$gene_biotype == \"protein_coding\", ]$associated_gene)),\n",
    "                      group = \"Isoform Atlas\") %>%\n",
    "        as.data.frame() %>%\n",
    "        unique(),\n",
    "      gencode %>%\n",
    "        # dplyr::filter(transcript_type == \"protein_coding\") %>%\n",
    "        dplyr::group_by(associated_gene) %>%\n",
    "        dplyr::summarise(count = dplyr::n()) %>%\n",
    "        dplyr::arrange(count) %>%\n",
    "        as.data.frame() %>%\n",
    "        dplyr::group_by(count) %>%\n",
    "        dplyr::summarise(isoform_count = dplyr::n()) %>%\n",
    "        as.data.frame() %>%\n",
    "        dplyr::mutate(count = ifelse(count>=10, \"10<=\", count),\n",
    "                     count = factor(count,levels = c(\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10<=\"))) %>%\n",
    "        dplyr::group_by(count) %>%\n",
    "        dplyr::mutate(isoform_count = sum(isoform_count),\n",
    "                      proportion = 100*isoform_count/length(unique(gencode %>%\n",
    "                                                                      .$associated_gene)),\n",
    "                      group = \"GENCOD38\") %>%\n",
    "        as.data.frame() %>%\n",
    "        unique()\n",
    ") %>%\n",
    "  ggplot(., aes(x = count, y = proportion, fill = group)) + \n",
    "  geom_bar(position=\"dodge\", stat=\"identity\") +\n",
    "  xlab(paste0(\"number of isoforms per gene\")) + \n",
    "  ylab(paste0(\"proportion(%)\")) +\n",
    "  theme_classic() +\n",
    "  theme(strip.text.x=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        strip.text.y=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=20),\n",
    "        axis.title.x = element_text(size=15),\n",
    "        axis.title.y = element_text(size =15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        axis.text.x = element_text(size = 15),\n",
    "        legend.text =  element_text(size = 10), \n",
    "        legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "        legend.title = element_text(size = 0, hjust = 0))\n",
    "\n",
    "fig1d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "036e9041-4f06-4dfd-bfa9-0e3b6df30ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=3)\n",
    "\n",
    "# structural category by SQANTI3 in comparison with GENCODE38\n",
    "out %>%\n",
    "  dplyr::group_by(structural_category_SQANTI3) %>%\n",
    "  dplyr::summarise(count = dplyr::n()) %>%\n",
    "  dplyr::mutate(structural_category_SQANTI3 = case_when(\n",
    "    structural_category_SQANTI3 == \"full-splice_match\" ~ \"FSM\",\n",
    "    structural_category_SQANTI3 == \"incomplete-splice_match\" ~ \"ISM\",\n",
    "    structural_category_SQANTI3 == \"novel_in_catalog\" ~ \"NIC\",\n",
    "    structural_category_SQANTI3 == \"novel_not_in_catalog\" ~ \"NNC\",\n",
    "    structural_category_SQANTI3 == \"intergenic\" ~ \"Intergenic\",\n",
    "    structural_category_SQANTI3 == \"antisense\" ~ \"Antisense\",\n",
    "    structural_category_SQANTI3 == \"fusion\" ~ \"Read-through\",\n",
    "    structural_category_SQANTI3 == \"genic\" ~ \"Genic\",\n",
    "    TRUE ~ structural_category_SQANTI3\n",
    "  ),\n",
    "               structural_category_SQANTI3 = forcats::fct_rev(factor(structural_category_SQANTI3,levels=c(\"FSM\",\"ISM\",\"NIC\",\"NNC\",\"Read-through\",\"Genic\",\"Intergenic\",\"Antisense\")))) %>%\n",
    "  dplyr::mutate(proportion = 100*count/sum(count)) %>%\n",
    "  dplyr::arrange(proportion)\n",
    "\n",
    "fig1e = out %>%\n",
    "  dplyr::group_by(structural_category_SQANTI3) %>%\n",
    "  dplyr::summarise(count = dplyr::n()) %>%\n",
    "  dplyr::mutate(structural_category_SQANTI3 = case_when(\n",
    "    structural_category_SQANTI3 == \"full-splice_match\" ~ \"FSM\",\n",
    "    structural_category_SQANTI3 == \"incomplete-splice_match\" ~ \"ISM\",\n",
    "    structural_category_SQANTI3 == \"novel_in_catalog\" ~ \"NIC\",\n",
    "    structural_category_SQANTI3 == \"novel_not_in_catalog\" ~ \"NNC\",\n",
    "    structural_category_SQANTI3 == \"intergenic\" ~ \"Intergenic\",\n",
    "    structural_category_SQANTI3 == \"antisense\" ~ \"Antisense\",\n",
    "    structural_category_SQANTI3 == \"fusion\" ~ \"Read-through\",\n",
    "    structural_category_SQANTI3 == \"genic\" ~ \"Genic\",\n",
    "    TRUE ~ structural_category_SQANTI3\n",
    "  ),\n",
    "               structural_category_SQANTI3 = forcats::fct_rev(factor(structural_category_SQANTI3,levels=c(\"FSM\",\"ISM\",\"NIC\",\"NNC\",\"Read-through\",\"Genic\",\"Intergenic\",\"Antisense\")))) %>%\n",
    "  dplyr::mutate(proportion = 100*count/sum(count)) %>%\n",
    "  dplyr::arrange(proportion) %>%\n",
    "  as.data.frame() %>%\n",
    "  ggplot(., aes(x = structural_category_SQANTI3, y = proportion, fill = structural_category_SQANTI3)) + \n",
    "  geom_bar(position=\"dodge\", stat=\"identity\") +\n",
    "  scale_fill_manual(values=c(\"FSM\" = \"darkblue\",\"ISM\" = \"orange\",\"NIC\" = \"green\",\"NNC\" = \"red\",\"Read-through\" = \"turquoise\",\"Genic\" = \"purple\",\"Intergenic\" = \"gold\" ,\"Antisense\" = \"grey\")) +\n",
    "  xlab(paste0(\"\")) + \n",
    "  ylab(paste0(\"Proportion(%)\")) +\n",
    "  coord_flip() +\n",
    "  theme_classic() +\n",
    "  theme(strip.text.x=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        strip.text.y=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        legend.position = \"none\",\n",
    "        plot.title = element_text(size=20),\n",
    "        axis.title.x = element_text(size=15),\n",
    "        axis.title.y = element_text(size =15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        axis.text.x = element_text(size = 15),\n",
    "        legend.text =  element_text(size = 15), \n",
    "        legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "        legend.title = element_text(size = 0, hjust = 0))\n",
    "\n",
    "\n",
    "fig1e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "126a0c95-5986-49e4-96f2-781d1ade3216",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=4, repr.plot.height=3)\n",
    "\n",
    "# splicing events by SUPPA2\n",
    "fig1f = rbind(\n",
    "  lr_suppa %>%\n",
    "    # dplyr::filter(alternative_transcripts %in% squanti[squanti$min_cov>0,]$isoform) %>%  # min_cov: minimum junction coverage based on short read STAR junction output file. NA if no short read given.\n",
    "    tidyr::separate_rows(.,alternative_transcripts, sep = \",\", convert = TRUE) %>%\n",
    "    dplyr::mutate(event_id = stringr::str_split(event_id,pattern=\";\",simplify=TRUE) %>% .[,2] %>% stringr::str_split(.,pattern=\":\",simplify=TRUE) %>% .[,1]) %>%\n",
    "    dplyr::select(c(event_id,alternative_transcripts)) %>%\n",
    "    dplyr::group_by(event_id) %>%\n",
    "    dplyr::summarise(count = n()) %>%\n",
    "    dplyr::filter(event_id!=\"\") %>%\n",
    "    dplyr::mutate(proportion = 100*count/sum(count),\n",
    "                  group = \"Isoform Atlas\",\n",
    "                  event_id = factor(event_id, levels = c(\"SE\",\"MX\",\"A5\",\"A3\",\"RI\",\"AF\",\"AL\"))),\n",
    "  gen_suppa %>%\n",
    "    tidyr::separate_rows(.,alternative_transcripts, sep = \",\", convert = TRUE) %>%\n",
    "    dplyr::mutate(event_id = stringr::str_split(event_id,pattern=\";\",simplify=TRUE) %>% .[,2] %>% stringr::str_split(.,pattern=\":\",simplify=TRUE) %>% .[,1]) %>%\n",
    "    dplyr::select(c(event_id,alternative_transcripts)) %>%\n",
    "    dplyr::group_by(event_id) %>%\n",
    "    dplyr::summarise(count = n()) %>%\n",
    "    dplyr::filter(event_id!=\"\") %>%\n",
    "    dplyr::mutate(proportion = 100*count/sum(count),\n",
    "                  group = \"GENCODE38\",\n",
    "                  event_id = factor(event_id, levels = c(\"SE\",\"MX\",\"A5\",\"A3\",\"RI\",\"AF\",\"AL\")))\n",
    ") %>%\n",
    "  # dplyr::group_by(event_id) %>%\n",
    "  # dplyr::mutate(diff = max(proportion)-min(proportion)) %>%\n",
    "  # as.data.frame() %>%\n",
    "  # dplyr::arrange(-diff) %>%\n",
    "  # dplyr::mutate(event_id = factor(event_id,levels=unique(event_id))) %>%\n",
    "  ggplot(data=., aes(x=event_id, y=proportion, fill=group)) +\n",
    "  geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "  theme_classic() +\n",
    "  # ggtitle (\"\") +\n",
    "  ylab(\"proportion(%)\") +\n",
    "  xlab(\"splicing events\") +\n",
    "  theme(strip.text.x=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        strip.text.y=element_text(size=20, color=\"black\", face=\"bold\"),\n",
    "        legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=20),\n",
    "        axis.title.x = element_text(size=15),\n",
    "        axis.title.y = element_text(size =15),\n",
    "        axis.text.y = element_text(size = 15),\n",
    "        axis.text.x = element_text(size = 15),\n",
    "        legend.text =  element_text(size = 10), \n",
    "        legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "        legend.title = element_text(size = 0, hjust = 0))\n",
    "\n",
    "fig1f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f174c953-e024-4d6a-ae95-4d8a7702795d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=6, repr.plot.height=4)\n",
    "fig1g = rbind(AS_susceptibility_pathway %>%\n",
    "  dplyr::filter(`Adjusted P-value` < 0.05) %>%\n",
    "  dplyr::mutate(logFDR = -log10(`Adjusted P-value`),\n",
    "                label = ifelse(`Adjusted P-value`<0.001, \"***\", ifelse(`Adjusted P-value`<0.01, \"**\", ifelse(`Adjusted P-value`<0.05, \"*\", \"\"))),\n",
    "                group = \"susceptible\"),\n",
    "                AS_constrained_pathway %>%\n",
    "  dplyr::filter(`Adjusted P-value` < 0.05) %>%\n",
    "  dplyr::mutate(logFDR = -log10(`Adjusted P-value`),\n",
    "                label = ifelse(`Adjusted P-value`<0.001, \"***\", ifelse(`Adjusted P-value`<0.01, \"**\", ifelse(`Adjusted P-value`<0.05, \"*\", \"\"))),\n",
    "               group = \"constrained\")) %>%\n",
    "  ggplot(., aes(x = reorder(x = Term, X = logFDR, FUN = sum), y = logFDR, fill = group)) + \n",
    "  geom_bar(stat = \"identity\") +\n",
    "  ylab(\"-log(FDR)\") +\n",
    "  xlab(\"\") +\n",
    "  # ggtitle(\"pathway analysis of AS susceptible genes\") +\n",
    "  coord_flip() +\n",
    "  theme_classic() +\n",
    "  scale_fill_manual(values=c(\"susceptible\" = \"purple4\", \"constrained\" = \"lightgoldenrod4\")) +\n",
    "  ggh4x::facet_nested(factor(group,levels=c(\"susceptible\",\"constrained\")) ~ ., scales = \"free\", nest_line = TRUE) +\n",
    "  ggh4x::force_panelsizes(rows = c(6, 1)) +\n",
    "  theme(strip.background = element_blank(),\n",
    "        strip.text.x=element_blank(),\n",
    "        strip.text.y=element_blank(),\n",
    "        legend.position = \"bottom\",\n",
    "        plot.title = element_text(size=20),\n",
    "        axis.title.x = element_text(size=12),\n",
    "        axis.title.y = element_text(size =12),\n",
    "        axis.text.y = element_text(size = 12),\n",
    "        axis.text.x = element_text(size = 12),\n",
    "        legend.text =  element_text(size = 12), \n",
    "        legend.key.size = grid::unit(0.8, \"lines\"),\n",
    "        legend.title = element_text(size = 0, hjust = 0))\n",
    "fig1g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50c3dcc9-7514-4fd6-b627-f21d0d5cef31",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width=8.5, repr.plot.height=11)\n",
    "ggarrange(ggplot() + theme_void(),\n",
    "          ggarrange(fig1b,g1,g3, ncol = 3, labels = c(\"B\", \"C\", \"\"), align = \"h\",widths = c(1,1,1),font.label = list(size = 15, color = \"black\")), \n",
    "          ggarrange(fig1d,　ggarrange(ggplot() + theme_void(), fig1e, ncol = 1, labels = c(\"\", \"\"),heights = c(0.4,1),font.label = list(size = 15, color = \"black\")), ncol = 2, labels = c(\"D\", \"E\"), align = \"h\",widths = c(1,1),font.label = list(size = 15, color = \"black\")), \n",
    "          ggarrange(fig1f,fig1g, ncol = 2, labels = c(\"F\", \"G\"), align = \"h\",widths = c(1,1.3),font.label = list(size = 15, color = \"black\")), \n",
    "          nrow = 4, \n",
    "          heights = c(1,1,1,1),\n",
    "          labels = \"A\",\n",
    "          font.label = list(size = 15, color = \"black\")\n",
    "          ) \n",
    "\n",
    "\n",
    "pdf(\"./Figure1.pdf\", width=8.5, height=11)\n",
    "ggarrange(ggplot() + theme_void(),\n",
    "          ggarrange(fig1b,g1,g3, ncol = 3, labels = c(\"B\", \"C\", \"\"), align = \"h\",widths = c(1,1,1),font.label = list(size = 15, color = \"black\")), \n",
    "          ggarrange(fig1d,　ggarrange(ggplot() + theme_void(), fig1e, ncol = 1, labels = c(\"\", \"\"),heights = c(0.4,1),font.label = list(size = 15, color = \"black\")), ncol = 2, labels = c(\"D\", \"E\"), align = \"h\",widths = c(1,1),font.label = list(size = 15, color = \"black\")), \n",
    "          ggarrange(fig1f,fig1g, ncol = 2, labels = c(\"F\", \"G\"), align = \"h\",widths = c(1,1.3),font.label = list(size = 15, color = \"black\")), \n",
    "          nrow = 4, \n",
    "          heights = c(1,1,1,1),\n",
    "          labels = \"A\",\n",
    "          font.label = list(size = 15, color = \"black\")\n",
    "          ) \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d29cfa5-dfe4-4f10-b647-8615e9513e03",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
